File: src/bumperbot_bridges/bumperbot_bridges/camera_stream.py
import rclpy
from rclpy.node import Node
from sensor_msgs.msg import Image, CameraInfo, LaserScan, PointCloud2
from cv_bridge import CvBridge
import cv2

class CameraStream(Node):
    def __init__(self):
        super().__init__('camera_stream')
        self.sub = self.create_subscription(Image, '/camera1/image/image', self.cb, 100)
        self.depth_sub = self.create_subscription(Image, '/camera1/image/depth_image', self.dcb, 100)
        self.info_sub = self.create_subscription(CameraInfo, '/camera1/image/camera_info', self.icb, 100)
        self.lidar_sub = self.create_subscription(LaserScan, '/lidar/cloud', self.lidarcb, 100)
        self.lidar_3D_sub = self.create_subscription(PointCloud2, '/lidar/cloud/points', self.lidar3Dcb, 100)

        self.pub = self.create_publisher(Image, '/camera1/preprocessed_out', 100)
        self.repub = self.create_publisher(Image, '/camera1/image', 100)
        self.depth_repub = self.create_publisher(Image, '/camera1/depth_image', 100)
        self.info_repub = self.create_publisher(CameraInfo, '/camera1/camera_info', 100)
        self.lidar_repub = self.create_publisher(LaserScan, '/lidar/scan', 100)
        self.lidar_3D_repub = self.create_publisher(PointCloud2, '/lidar/scan/points', 100)
        self.br = CvBridge()

    def cb(self, msg):
        self.fixImageFrame(msg)
        cv_img = self.br.imgmsg_to_cv2(msg, 'bgr8')
        cv_img = self.preprocess(cv_img)
        new_msg = self.br.cv2_to_imgmsg(cv_img, 'bgr8')
        self.pub.publish(new_msg)

    def lidarcb(self, msg):
        msg.header.frame_id = 'lidar_link'
        self.lidar_repub.publish(msg)

    def lidar3Dcb(self, msg):
        msg.header.frame_id = 'lidar_link'
        self.lidar_3D_repub.publish(msg)

    def dcb(self, msg):
        msg.header.frame_id = 'camera_link'
        self.depth_repub.publish(msg)

    def icb(self, msg):
        msg.header.frame_id = 'camera_link'
        self.info_repub.publish(msg)

    def fixImageFrame(self, msg):
        msg.header.frame_id = 'camera_link'
        self.repub.publish(msg)


    def preprocess(self, img):
        cv_img = cv2.resize(img, (320, 320))
        return cv_img
    

def main():
    rclpy.init()
    node = CameraStream()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()

File: src/cpp_bridges/cpp_bridges/camera_stream.cpp
#include "rclcpp/rclcpp.hpp"
#include "sensor_msgs/msg/image.hpp"
#include "sensor_msgs/msg/camera_info.hpp"
#include "sensor_msgs/msg/laser_scan.hpp"
#include "sensor_msgs/msg/point_cloud2.hpp"
#include "cv_bridge/cv_bridge.hpp"
#include <opencv2/opencv.hpp>

class CameraStream : public rclcpp::Node {
public:
    CameraStream() : Node("camera_stream") {
        // Subscriptions
        sub_ = this->create_subscription<sensor_msgs::msg::Image>(
            "/camera1/image/image", 10,
            std::bind(&CameraStream::cb, this, std::placeholders::_1));

        depth_sub_ = this->create_subscription<sensor_msgs::msg::Image>(
            "/camera1/image/depth_image", 10,
            std::bind(&CameraStream::dcb, this, std::placeholders::_1));

        info_sub_ = this->create_subscription<sensor_msgs::msg::CameraInfo>(
            "/camera1/image/camera_info", 10,
            std::bind(&CameraStream::icb, this, std::placeholders::_1));

        lidar_sub_ = this->create_subscription<sensor_msgs::msg::LaserScan>(
            "/lidar/cloud", 10,
            std::bind(&CameraStream::lidarcb, this, std::placeholders::_1));

        lidar3d_sub_ = this->create_subscription<sensor_msgs::msg::PointCloud2>(
            "/lidar/cloud/points", 10,
            std::bind(&CameraStream::lidar3Dcb, this, std::placeholders::_1));

        // Publishers
        pub_ = this->create_publisher<sensor_msgs::msg::Image>("/camera1/preprocessed_out", 10);
        repub_ = this->create_publisher<sensor_msgs::msg::Image>("/camera1/image", 10);
        depth_repub_ = this->create_publisher<sensor_msgs::msg::Image>("/camera1/depth_image", 10);
        info_repub_ = this->create_publisher<sensor_msgs::msg::CameraInfo>("/camera1/camera_info", 10);
        lidar_repub_ = this->create_publisher<sensor_msgs::msg::LaserScan>("/lidar/scan", 10);
        lidar3d_repub_ = this->create_publisher<sensor_msgs::msg::PointCloud2>("/lidar/scan/points", 10);
    }

private:
    // ========== Image callbacks ==========
    void cb(const sensor_msgs::msg::Image::ConstSharedPtr& msg) {
        auto fixed_msg = *msg;
        fixImageFrame(fixed_msg);

        // Convert to OpenCV
        cv_bridge::CvImagePtr cv_ptr;
        try {
            cv_ptr = cv_bridge::toCvCopy(msg, "bgr8");
        } catch (cv_bridge::Exception& e) {
            RCLCPP_ERROR(this->get_logger(), "cv_bridge exception: %s", e.what());
            return;
        }

        // Preprocess
        cv::Mat preprocessed = preprocess(cv_ptr->image);

        // Convert back to ROS message
        sensor_msgs::msg::Image::SharedPtr new_msg =
            cv_bridge::CvImage(msg->header, "bgr8", preprocessed).toImageMsg();

        pub_->publish(*new_msg);
    }

    void dcb(const sensor_msgs::msg::Image::ConstSharedPtr& msg) {
        auto new_msg = *msg;
        new_msg.header.frame_id = "camera_link";
        depth_repub_->publish(new_msg);
    }

    void icb(const sensor_msgs::msg::CameraInfo::ConstSharedPtr& msg) {
        auto new_msg = *msg;
        new_msg.header.frame_id = "camera_link";
        info_repub_->publish(new_msg);
    }

    void lidarcb(const sensor_msgs::msg::LaserScan::ConstSharedPtr& msg) {
        auto new_msg = *msg;
        new_msg.header.frame_id = "lidar_link";
        lidar_repub_->publish(new_msg);
    }

    void lidar3Dcb(const sensor_msgs::msg::PointCloud2::ConstSharedPtr& msg) {
        auto new_msg = *msg;
        new_msg.header.frame_id = "lidar_link";
        lidar3d_repub_->publish(new_msg);
    }

    // ========== Helper methods ==========
    void fixImageFrame(sensor_msgs::msg::Image& msg) {
        msg.header.frame_id = "camera_link";
        repub_->publish(msg);
    }

    cv::Mat preprocess(const cv::Mat& img) {
        cv::Mat resized;
        cv::resize(img, resized, cv::Size(320, 320));
        return resized;
    }

    // ========== Members ==========
    rclcpp::Subscription<sensor_msgs::msg::Image>::SharedPtr sub_, depth_sub_;
    rclcpp::Subscription<sensor_msgs::msg::CameraInfo>::SharedPtr info_sub_;
    rclcpp::Subscription<sensor_msgs::msg::LaserScan>::SharedPtr lidar_sub_;
    rclcpp::Subscription<sensor_msgs::msg::PointCloud2>::SharedPtr lidar3d_sub_;

    rclcpp::Publisher<sensor_msgs::msg::Image>::SharedPtr pub_, repub_, depth_repub_;
    rclcpp::Publisher<sensor_msgs::msg::CameraInfo>::SharedPtr info_repub_;
    rclcpp::Publisher<sensor_msgs::msg::LaserScan>::SharedPtr lidar_repub_;
    rclcpp::Publisher<sensor_msgs::msg::PointCloud2>::SharedPtr lidar3d_repub_;
};

int main(int argc, char* argv[]) {
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<CameraStream>());
    rclcpp::shutdown();
    return 0;
}
